File: src/main/java/com/kirinpatel/sync/gui/ClientInfoGUI.java
package com.kirinpatel.sync.gui;

import com.kirinpatel.sync.net.Server;
import com.kirinpatel.sync.net.User;

import javax.swing.*;
import java.awt.*;
import java.awt.event.ComponentEvent;
import java.awt.event.ComponentListener;

class ClientInfoGUI extends JFrame {

    private final User user;
    private final JLabel ping;
    private final JLabel mediaTime;
    private final JLabel mediaState;
    private final UpdateUIThread updateUIThread;

    ClientInfoGUI(User user) {
        super("Client info");

        this.user = user;
        setResizable(false);
        setLayout(new GridLayout(6, 1));
        setDefaultCloseOperation(HIDE_ON_CLOSE);
        addComponentListener(new ClientInfoComponentListener());
        setLocationRelativeTo(null);

        add(new JLabel("Username: " + user.getUsername()));
        add(new JLabel("UserID: " + user.getUserID()));
        ping = new JLabel("Ping: " + user.getPing() + " ms");
        add(ping);
        mediaTime = new JLabel("Playback Time: "
                + user.getMedia().getFormattedTime()
                + ":" + user.getMedia().getCurrentTime() % 1000);
        add(mediaTime);
        mediaState = new JLabel("Playback State: "
                + (user.getMedia().isPaused() ? "Paused" : "Playing"));
        add(mediaState);
        JButton disconnectUser = new JButton("Kick Client");
        updateUIThread = new UpdateUIThread();
        disconnectUser.addActionListener(e -> {
            Server.kickUser(user);
            ControlPanel.isUserDisplayShown = false;
            updateUIThread.stop();
            dispose();
        });
        add(disconnectUser);
        pack();

        new Thread(updateUIThread).start();

        setVisible(true);
    }

    class ClientInfoComponentListener implements ComponentListener {

        @Override
        public void componentResized(ComponentEvent e) {

        }

        @Override
        public void componentMoved(ComponentEvent e) {

        }

        @Override
        public void componentShown(ComponentEvent e) {

        }

        @Override
        public void componentHidden(ComponentEvent e) {
            ControlPanel.isUserDisplayShown = false;
            updateUIThread.stop();
            dispose();
        }
    }

    class UpdateUIThread implements Runnable {

        private boolean isRunning = true;

        @Override
        public void run() {
            while(isRunning) {
                try {
                    Thread.sleep(250);
                    ping.setText("Ping: " + user.getPing() + " ms");
                    mediaTime.setText("Playback Time: "
                            + user.getMedia().getFormattedTime()
                            + ":" + user.getMedia().getCurrentTime() % 1000);
                    mediaState.setText("Playback State: "
                            + (user.getMedia().isPaused() ? "Paused" : "Playing"));
                } catch(InterruptedException e) {
                    Thread.currentThread().interrupt();
                }
            }
        }

        void stop() {
            isRunning = false;
        }
    }
}

File: src/main/java/com/kirinpatel/sync/gui/ClientInfoGUI.kt
package com.kirinpatel.sync.gui

import com.kirinpatel.sync.net.Server
import com.kirinpatel.sync.net.User
import java.awt.GridLayout
import java.awt.event.ComponentEvent
import java.awt.event.ComponentListener
import javax.swing.JButton
import javax.swing.JFrame
import javax.swing.JLabel
import javax.swing.WindowConstants

class ClientInfoGUI constructor(val user: User): JFrame("Client info") {
    val ping = JLabel("Ping: " + user.ping + " ms")
    val mediaState = JLabel("Playback State: " + if (user.media.isPaused) "Paused" else "Playing")
    val mediaTime = JLabel("Playback Time: "
            + user.media.formattedTime
            + ":" + user.media.currentTime % 1000)
    internal val updateUIThread = UpdateUIThread()

    init {
        isResizable = false
        layout = GridLayout(6, 1)
        defaultCloseOperation = WindowConstants.HIDE_ON_CLOSE
        addComponentListener(ClientInfoComponentListener())
        setLocationRelativeTo(null)

        add(JLabel("Username: " + user.username))
        add(JLabel("UserID: " + user.userID))
        add(ping)
        add(mediaTime)
        add(mediaState)
        val disconnectUser = JButton("Kick Client")
        disconnectUser.addActionListener { _ ->
            Server.kickUser(user)
            ControlPanel.isUserDisplayShown = false
            updateUIThread.stop()
            dispose()
        }
        add(disconnectUser)
        pack()

        Thread(updateUIThread).start()

        isVisible = true
    }

    internal inner class ClientInfoComponentListener : ComponentListener {

        override fun componentResized(e: ComponentEvent) {

        }

        override fun componentMoved(e: ComponentEvent) {

        }

        override fun componentShown(e: ComponentEvent) {

        }

        override fun componentHidden(e: ComponentEvent) {
            ControlPanel.isUserDisplayShown = false
            updateUIThread.stop()
            dispose()
        }
    }

    internal inner class UpdateUIThread : Runnable {
        private var isRunning = true

        override fun run() {
            while (isRunning) {
                try {
                    Thread.sleep(250)
                    ping.text = "Ping: " + user.ping + " ms"
                    mediaTime.text = ("Playback Time: "
                            + user.media.formattedTime
                            + ":" + user.media.currentTime % 1000)
                    mediaState.text = "Playback State: " + if (user.media.isPaused) "Paused" else "Playing"
                } catch (e: InterruptedException) {
                    Thread.currentThread().interrupt()
                }

            }
        }

        fun stop() {
            isRunning = false
        }
    }
}